libs/entitlement/asm-provider/src/main/java/org/elasticsearch/entitlement/instrumentation/impl/InstrumentationServiceImpl.java

```diff
@@ -53,94 +55,91 @@ public Map<MethodKey, CheckerMethod> lookupMethodsToInstrument(String entitlemen
-                                methodDescriptor,
-                                methodsToInstrument
-                            );
-                        }
-                        return av;
-                    }
-                };
+            public MethodVisitor visitMethod(
+                int access,
+                String checkerMethodName,
+                String checkerMethodDescriptor,
+                String signature,
+                String[] exceptions
+            ) {
+                var mv = super.visitMethod(access, checkerMethodName, checkerMethodDescriptor, signature, exceptions);
+
+                var checkerMethodArgumentTypes = Type.getArgumentTypes(checkerMethodDescriptor);
+                var methodToInstrument = parseCheckerMethodSignature(checkerMethodName, checkerMethodArgumentTypes);
+
+                var checkerParameterDescriptors = Arrays.stream(checkerMethodArgumentTypes).map(Type::getDescriptor).toList();
+                var checkerMethod = new CheckerMethod(Type.getInternalName(checkerClass), checkerMethodName, checkerParameterDescriptors);
+
+                methodsToInstrument.put(methodToInstrument, checkerMethod);
+
+                return mv;
             }
         };
         reader.accept(visitor, 0);
         return methodsToInstrument;
     }
 
-    private static class InstrumentationTargetAnnotationVisitor extends AnnotationVisitor {
-        private final String checkerClassName;
-        private final String checkerMethodName;
-        private final String checkerMethodDescriptor;
-        private final HashMap<MethodKey, CheckerMethod> methodsToInstrument;
-        private String targetClassName;
-        private String targetMethodName;
-        private boolean targetMethodIsStatic;
-
-        InstrumentationTargetAnnotationVisitor(
-            AnnotationVisitor av,
-            String checkerClassName,
-            String checkerMethodName,
-            String checkerMethodDescriptor,
-            HashMap<MethodKey, CheckerMethod> methodsToInstrument
-        ) {
-            super(Opcodes.ASM9, av);
-            this.checkerClassName = checkerClassName;
-            this.checkerMethodName = checkerMethodName;
-            this.checkerMethodDescriptor = checkerMethodDescriptor;
-            this.methodsToInstrument = methodsToInstrument;
+    private static final Type CLASS_TYPE = Type.getType(Class.class);
+
+    static MethodKey parseCheckerMethodSignature(String checkerMethodName, Type[] checkerMethodArgumentTypes) {
+
+        var methodNameTokens = checkerMethodName.split("\\$");
+        if (methodNameTokens.length < 3 || "check".equals(methodNameTokens[0]) == false) {
+            throw new IllegalArgumentException(
+                String.format(
+                    Locale.ROOT,
+                    "Checker method %s has incorrect name format. "
+                        + "It should be either check$$methodName (instance) or check$package_ClassName$methodName (static)",
```

Comment:
Huh, that's kind of clever, but I'm torn on it. On the one hand, this reduces redundancy because the class could already be derived from the argument list in the case of an instance method. But on the other hand, a uniform name-mangling convention seems desirable too. 🤔